File: app/src/main/java/xyz/projectplay/jisho/ui/recyclerview/WordViewHolder.java
/*
 * Copyright 2016 Nathaniel Reline
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package xyz.projectplay.jisho.ui.recyclerview;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import xyz.projectplay.jisho.R;
import xyz.projectplay.jisho.models.Japanese;
import xyz.projectplay.jisho.models.Sense;
import xyz.projectplay.jisho.models.Word;
import xyz.projectplay.jisho.ui.widgets.AutoResizeTextView;

public class WordViewHolder extends RecyclerView.ViewHolder {

    @BindView(R.id.word)
    AutoResizeTextView readingTextView;

    @BindView(R.id.furigana)
    TextView furiganaTextView;

    @BindView(R.id.common)
    TextView commonTextView;

    @BindView(R.id.tags)
    LinearLayout tagsLayout;

    @BindView(R.id.senses)
    LinearLayout sensesLayout;

    public WordViewHolder(@NonNull View itemView) {
        super(itemView);
        ButterKnife.bind(this, itemView);
    }

    public void bind(Word word) {
        Japanese japanese = word.getJapanese().get(0); // always use the first result
        if (japanese.getWord() != null) {
            furiganaTextView.setVisibility(View.VISIBLE);
            furiganaTextView.setText(japanese.getReading());
            readingTextView.setText(japanese.getWord());
        } else {
            furiganaTextView.setVisibility(View.GONE);
            readingTextView.setText(japanese.getReading());
        }

        commonTextView.setVisibility(word.isCommon() ? View.VISIBLE : View.GONE);

        tagsLayout.removeAllViews();
        for (String tag : word.getTags()) {
            TextView textView = (TextView) View.inflate(itemView.getContext(), R.layout.layout_tag, null);
            textView.setText(tag);
            tagsLayout.addView(textView);
        }

        bindSenses(word.getSenses());
    }

    private void bindSenses(List<Sense> senses) {
        sensesLayout.removeAllViews();
        for (int i = 0; i < senses.size(); i++) {
            Sense sense = senses.get(i);
            TextView textView = new TextView(itemView.getContext());
            String definition = String.valueOf(i + 1) + ". ";
            List<String> englishDefinitions = sense.getEnglishDefinitions();
            for (int k = 0; k < englishDefinitions.size(); k++) {
                definition += englishDefinitions.get(k) + (k < englishDefinitions.size() - 1 ? "; " : "");
            }
            textView.setText(definition);
            textView.setTextIsSelectable(true);
            sensesLayout.addView(textView);
        }
    }
}

File: android/src/main/java/com/github/reline/jisho/words/WordViewHolder.kt
/*
 * Copyright 2020 Nathaniel Reline
 *
 * This work is licensed under the Creative Commons Attribution-ShareAlike 4.0 International License.
 * To view a copy of this license, visit http://creativecommons.org/licenses/by-sa/4.0/ or
 * send a letter to Creative Commons, PO Box 1866, Mountain View, CA 94042, USA.
 */

package com.github.reline.jisho.words

import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import android.view.View
import android.widget.TextView
import com.github.reline.jisho.R
import com.github.reline.jisho.databinding.CardWordBinding
import com.github.reline.jisho.models.Definition
import com.github.reline.jisho.models.Result

class WordViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

    fun bind(result: Result) {
        with(CardWordBinding.bind(itemView)) {
            if (result.okurigana != null) {
                cardFuriganaTextView.visibility = View.VISIBLE
                cardFuriganaTextView.text = result.okurigana
            } else {
                cardFuriganaTextView.visibility = View.GONE
            }
            cardReadingTextView.text = result.japanese

            cardCommonTextView.visibility = if (result.isCommon) View.VISIBLE else View.GONE

            bindTags(result.tags)
            bindSenses(result.senses)
        }
    }

    private fun bindTags(tags: List<String>) {
        with(CardWordBinding.bind(itemView)) {
            cardTagsLayout.removeAllViews()
            for (tag in tags) {
                val textView = View.inflate(root.context, R.layout.layout_tag, null) as TextView
                textView.text = tag
                cardTagsLayout.addView(textView)
            }
        }
    }

    private fun bindSenses(senses: List<Definition>) {
        with(CardWordBinding.bind(itemView)) {
            cardSensesLayout.removeAllViews()
            for (i in senses.indices) {
                val sense = senses[i]

                bindPartsOfSpeech(sense.partsOfSpeech)

                val definitionTextView = TextView(root.context)
                val definition = StringBuilder().append(i + 1).append(" ")
                val englishDefinitions = sense.values
                for (k in englishDefinitions.indices) {
                    val separator = if (k < englishDefinitions.size - 1) "; " else ""
                    definition.append(englishDefinitions[k]).append(separator)
                }
                definitionTextView.text = definition
                definitionTextView.setTextIsSelectable(true)
                cardSensesLayout.addView(definitionTextView)
            }
        }
    }

    private fun bindPartsOfSpeech(pos: List<String>) {
        with(CardWordBinding.bind(itemView)) {
            if (pos.isEmpty()) return
            val posTextView = TextView(root.context)
            val partsOfSpeech = StringBuilder()
            for (i in pos.indices) {
                val separator = if (i < pos.size - 1) ", " else ""
                partsOfSpeech.append(pos[i]).append(separator)
            }
            posTextView.text = partsOfSpeech.toString()
            posTextView.setTextColor(ContextCompat.getColor(root.context, R.color.colorAccent))
            posTextView.setTextIsSelectable(true)
            cardSensesLayout.addView(posTextView)
        }
    }
}
